File: services/catalog/src/domain/models/Money.ts
/**
 * Money Value Object
 *
 * Represents a monetary amount with currency, following the common.v1.Money
 * protobuf definition. Precision-safe using integer units + nanos pattern.
 *
 * Domain core has ZERO external dependencies.
 */
export class Money {
  public readonly currencyCode: string;
  public readonly units: number;
  public readonly nanos: number;

  private constructor(currencyCode: string, units: number, nanos: number) {
    this.currencyCode = currencyCode;
    this.units = units;
    this.nanos = nanos;
  }

  /**
   * Create a Money value object with validation.
   * Nanos must be same sign as units and in range [0, 1_000_000_000).
   */
  public static create(currencyCode: string, units: number, nanos: number): Money {
    if (!currencyCode || currencyCode.length !== 3) {
      throw new MoneyValidationError(
        `Invalid currency code: "${currencyCode}". Must be a 3-letter ISO 4217 code.`
      );
    }

    if (!Number.isInteger(units)) {
      throw new MoneyValidationError(
        `Units must be an integer, got: ${units}`
      );
    }

    if (!Number.isInteger(nanos)) {
      throw new MoneyValidationError(
        `Nanos must be an integer, got: ${nanos}`
      );
    }

    if (nanos <= -1_000_000_000 || nanos >= 1_000_000_000) {
      throw new MoneyValidationError(
        `Nanos must be in range [-999999999, 999999999], got: ${nanos}`
      );
    }

    // Sign consistency: nanos must be same sign as units (or zero)
    if (units > 0 && nanos < 0) {
      throw new MoneyValidationError(
        'Nanos must have the same sign as units (both positive or both negative).'
      );
    }
    if (units < 0 && nanos > 0) {
      throw new MoneyValidationError(
        'Nanos must have the same sign as units (both positive or both negative).'
      );
    }

    return new Money(currencyCode, units, nanos);
  }

  /**
   * Create Money from a decimal string, e.g. "29.99" with currency "USD".
   */
  public static fromDecimal(currencyCode: string, amount: string): Money {
    const parts = amount.split('.');
    const wholePart = parseInt(parts[0], 10);
    const decimalStr = parts[1] ?? '0';
    const paddedDecimal = decimalStr.padEnd(9, '0').slice(0, 9);
    let nanos = parseInt(paddedDecimal, 10);

    if (wholePart < 0) {
      nanos = -nanos;
    }

    return Money.create(currencyCode, wholePart, nanos);
  }

  /**
   * Create a zero-value Money in the given currency.
   */
  public static zero(currencyCode: string): Money {
    return new Money(currencyCode, 0, 0);
  }

  /**
   * Convert to a decimal string representation.
   */
  public toDecimal(): string {
    const sign = this.units < 0 || this.nanos < 0 ? '-' : '';
    const absUnits = Math.abs(this.units);
    const absNanos = Math.abs(this.nanos);
    const nanosStr = absNanos.toString().padStart(9, '0').replace(/0+$/, '');
    if (nanosStr.length === 0) {
      return `${sign}${absUnits}`;
    }
    return `${sign}${absUnits}.${nanosStr}`;
  }

  /**
   * Add two Money values of the same currency.
   */
  public add(other: Money): Money {
    this.assertSameCurrency(other);
    const totalNanos = this.nanos + other.nanos;
    const carry = Math.trunc(totalNanos / 1_000_000_000);
    const remainderNanos = totalNanos % 1_000_000_000;
    const resultUnits = this.units + other.units + carry;
    return new Money(this.currencyCode, resultUnits, remainderNanos);
  }

  /**
   * Subtract another Money value of the same currency.
   */
  public subtract(other: Money): Money {
    this.assertSameCurrency(other);
    const negated = new Money(other.currencyCode, -other.units, -other.nanos);
    return this.add(negated);
  }

  /**
   * Check if this Money is greater than another.
   */
  public greaterThan(other: Money): boolean {
    this.assertSameCurrency(other);
    if (this.units !== other.units) {
      return this.units > other.units;
    }
    return this.nanos > other.nanos;
  }

  /**
   * Check if this Money is less than another.
   */
  public lessThan(other: Money): boolean {
    this.assertSameCurrency(other);
    if (this.units !== other.units) {
      return this.units < other.units;
    }
    return this.nanos < other.nanos;
  }

  /**
   * Check equality with another Money value.
   */
  public equals(other: Money): boolean {
    return (
      this.currencyCode === other.currencyCode &&
      this.units === other.units &&
      this.nanos === other.nanos
    );
  }

  /**
   * Check if the amount is zero.
   */
  public isZero(): boolean {
    return this.units === 0 && this.nanos === 0;
  }

  /**
   * Check if the amount is negative.
   */
  public isNegative(): boolean {
    return this.units < 0 || (this.units === 0 && this.nanos < 0);
  }

  private assertSameCurrency(other: Money): void {
    if (this.currencyCode !== other.currencyCode) {
      throw new MoneyValidationError(
        `Currency mismatch: cannot operate on ${this.currencyCode} and ${other.currencyCode}`
      );
    }
  }
}

/**
 * Validation error for Money value object.
 */
export class MoneyValidationError extends Error {
  public readonly code = 'MONEY_VALIDATION_ERROR';

  constructor(message: string) {
    super(message);
    this.name = 'MoneyValidationError';
  }
}

File: services/order/src/main/kotlin/domain/models/Money.kt
package com.company.order.domain.models

/**
 * Value object representing a monetary amount with currency.
 * Follows ISO 4217 currency codes and provides precision-safe arithmetic.
 * Immutable by design — all operations return new instances.
 *
 * Domain core has ZERO external dependencies.
 */
data class Money(
    val units: Long,
    val nanos: Int,
    val currencyCode: String
) {
    companion object {
        /** The maximum absolute value for nanos (999,999,999) */
        private const val NANOS_PER_UNIT = 1_000_000_000L

        /** Create Money from a decimal representation: e.g. fromDecimal(29.99, "USD") */
        fun fromDecimal(amount: Double, currencyCode: String): Money {
            require(amount.isFinite()) { "Amount must be finite" }
            val sign = if (amount < 0) -1 else 1
            val absAmount = kotlin.math.abs(amount)
            val units = (absAmount.toLong()) * sign
            val nanos = ((absAmount - absAmount.toLong()) * NANOS_PER_UNIT).toInt() * sign
            return Money(units, nanos, currencyCode.uppercase())
        }

        /** Create Money from units and nanos with normalization */
        fun of(units: Long, nanos: Int, currencyCode: String): Money {
            val normalized = normalize(units, nanos)
            return Money(normalized.first, normalized.second, currencyCode.uppercase())
        }

        /** Zero amount in the given currency */
        fun zero(currencyCode: String): Money = Money(0, 0, currencyCode.uppercase())

        private fun normalize(units: Long, nanos: Int): Pair<Long, Int> {
            var u = units
            var n = nanos
            // Carry overflow from nanos to units
            if (n >= NANOS_PER_UNIT) {
                u += n / NANOS_PER_UNIT
                n = (n % NANOS_PER_UNIT).toInt()
            }
            // Ensure nanos and units have the same sign (or nanos is zero)
            if (u > 0 && n < 0) {
                u -= 1
                n = (n + NANOS_PER_UNIT).toInt()
            } else if (u < 0 && n > 0) {
                u += 1
                n = (n - NANOS_PER_UNIT).toInt()
            }
            return Pair(u, n)
        }
    }

    init {
        require(currencyCode.isNotBlank()) { "Currency code must not be blank" }
        require(currencyCode.length == 3) { "Currency code must be 3 characters (ISO 4217)" }
        require(kotlin.math.abs(nanos) < NANOS_PER_UNIT) {
            "Nanos absolute value must be < $NANOS_PER_UNIT, got $nanos"
        }
        require((units == 0L && nanos == 0) || (units > 0 && nanos >= 0) || (units < 0 && nanos <= 0)) {
            "Units and nanos must have the same sign (or one must be zero). units=$units, nanos=$nanos"
        }
    }

    val isNegative: Boolean get() = units < 0 || (units == 0L && nanos < 0)
    val isZero: Boolean get() = units == 0L && nanos == 0
    val isPositive: Boolean get() = units > 0 || (units == 0L && nanos > 0)

    /** Add two Money values. Must have the same currency. */
    operator fun plus(other: Money): Money {
        requireSameCurrency(other)
        return of(this.units + other.units, this.nanos + other.nanos, currencyCode)
    }

    /** Subtract two Money values. Must have the same currency. */
    operator fun minus(other: Money): Money {
        requireSameCurrency(other)
        return of(this.units - other.units, this.nanos - other.nanos, currencyCode)
    }

    /** Multiply by a scalar factor */
    fun multiply(factor: Int): Money = of(units * factor, nanos * factor, currencyCode)

    /** Negate the amount */
    fun negate(): Money = of(-units, -nanos, currencyCode)

    /** Absolute value */
    fun abs(): Money = if (isNegative) negate() else this

    private fun requireSameCurrency(other: Money) {
        require(currencyCode == other.currencyCode) {
            "Cannot operate on Money with different currencies: $currencyCode vs ${other.currencyCode}"
        }
    }

    override fun toString(): String = "${if (isNegative) "-" else ""}${kotlin.math.abs(units)}.${
        kotlin.math.abs(nanos).toString().padStart(9, '0').trimEnd('0')
    } $currencyCode"
}
